### PR TITLE
feat(rebaser): Rebaser uses layerdb activites

### PR DESCRIPTION
**Pull request by adamhjk**
This refactors the rebaser to use a layerdb activities stream. Under the hood it uses a NATS Jetstream work queue.

* Makes subscriptions to activities multiplexed (unless they are a work queue)
* Adds integration test activities
* Adds rebase and rebase_and_wait, which waits for a finish event

### CODE REVIEW

**Comment by adamhjk** [comment]
bors merge

File: lib/dal/src/context.rs [hunk 1]
```diff
@@ -1147,44 +1118,56 @@ pub struct Conflicts {
 // not want it long term.
 async fn rebase(
     tenancy: &Tenancy,
-    nats: NatsClient,
-    rebaser_config: RebaserClientConfig,
+    layer_db: &DalLayerDb,
     rebase_request: RebaseRequest,
 ) -> Result<Option<Conflicts>, TransactionsError> {
```
Comment (nickgerace) [inline lib/dal/src/context.rs:1123]: I dig it!

File: lib/rebaser-server/src/server/core_loop.rs [hunk 2]
```diff
@@ -61,33 +62,15 @@ pub(crate) async fn setup_and_run_core_loop(
         None,
         None,
         symmetric_crypto_service,
-        messaging_config.clone(),
-        layer_db,
+        layer_db.clone(),
     );
 
-    // Setup the subjects.
```
Comment (nickgerace) [inline lib/rebaser-server/src/server/core_loop.rs:68]: Splendid

File: lib/rebaser-server/src/server/core_loop.rs [hunk 3]
```diff
@@ -61,33 +62,15 @@ pub(crate) async fn setup_and_run_core_loop(
...
-        .await?;
-
-    info!(%subject_all, %subject_root, "getting stream from consumer");
-    let stream = consumer.stream().await?;
-
     info!("getting dal context builder");
     let ctx_builder = DalContext::builder(services_context.clone(), false);
 
+    info!("subscribing to work queue");
+    let stream = layer_db.activity().rebase().subscribe_work_queue().await?;
```
Comment (nickgerace) [inline lib/rebaser-server/src/server/core_loop.rs:72]: Drop-in replacement!

File: lib/rebaser-server/src/server/rebase.rs [hunk 4]
```diff
@@ -102,13 +108,13 @@ pub(crate) async fn perform_rebase(
...
+            updates_performed: serde_json::to_value(updates)?.to_string(),
         }
     } else {
-        ReplyRebaseMessage::ConflictsFound {
-            conflicts_found: serde_json::to_value(conflicts)?,
-            updates_found_and_skipped: serde_json::to_value(updates)?,
+        RebaseStatus::ConflictsFound {
+            conflicts_found: serde_json::to_value(conflicts)?.to_string(),
+            updates_found_and_skipped: serde_json::to_value(updates)?.to_string(),
         }
```
Comment (nickgerace) [inline lib/rebaser-server/src/server/rebase.rs:118]: For the future: potential use concrete types in the response. We might only need `Conflicts` too.

File: lib/si-layer-cache/src/activities/rebase.rs [hunk 5]
```diff
@@ -34,41 +43,143 @@ impl RebaseRequest {
...
+        &self.to_rebase_change_set_id
+    }
+
+    pub fn onto_workspace_snapshot_address(&self) -> &WorkspaceSnapshotAddress {
+        &self.onto_workspace_snapshot_address
+    }
 }
 
+// NOTE: We're basically smashing the data in here, and we really do have to figure out what we
+// actually want when things work / or don't work.
```
Comment (nickgerace) [inline lib/si-layer-cache/src/activities/rebase.rs:76]: Yep

File: lib/si-layer-cache/src/activities/rebase.rs [hunk 6]
```diff
@@ -34,41 +43,143 @@ impl RebaseRequest {
...
+            onto_workspace_snapshot_address,
+            onto_vector_clock_id,
+        );
+        let activity = Activity::rebase(payload, metadata);
+        debug!(?activity, "sending rebase and waiting for response");
+
+        // Why is this in two? We want to start listening before the publish call, to ensure we
+        // aren't racing with any listening service.
+        let start = Instant::now();
+        let rx = self.rebase_finished_activity_stream().await?;
```
Comment (nickgerace) [inline lib/si-layer-cache/src/activities/rebase.rs:146]: Good idea

**Review by nickgerace — APPROVED** [review]

**Pull request merged**